src/adaptors/adaptor_tls.h

```diff
@@ -0,0 +1,137 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "adaptor_buffer.h"
+#include "adaptor_common.h"
+
+#include "qpid/dispatch/alloc.h"
+#include "qpid/dispatch/protocol_adaptor.h"
+
+#include <proton/tls.h>
+
+#define QD_TLS_ERROR -1
+
+typedef struct qd_tls_t qd_tls_t;
+
+qd_tls_t *qd_tls(void *context, uint64_t conn_id, qd_log_source_t *log_source);
+
+/**
+ * Configure proton pn_tls objects using passed in config
+ * and start the tls session by calling the pn_tls_start function.
+ *     Logs sslProfile configuration failures at INFO level.
+ * On success:
+ *     pn_tls_config and pn_tls_session objects in tls are set up
+ *     Returns true
+ * On failure:
+ *     Error log is written
+ *     All in-progress pn_tls objects are destroyed.
+ *     Returns false.
+ *
+ * @param tls - Pointer to the qd_tls_t object which contains the tls session information
+ * @param config   - Pointer to the qd_adaptor_config_t object which contains the config information
+ * @param qd -     - Pointer to the qd_dispatch_t object.
+ * @param is_listener - set this flag to true if the tls session is initialized on a listener, false otherwise.
+ * @prototcols     - An array of supported TLS protocol version, TLSv1.2 or TSLv1.3 or both
+ */
+bool qd_tls_start(qd_tls_t                  *tls,
+                  const qd_adaptor_config_t *config,
+                  const qd_dispatch_t       *qd,
+                  bool                       is_listener,
+                  const char                *protocols[]);
+
+/**
+ * Takes as many read buffers from the raw connection that is allowed by pn_tls_get_decrypt_input_buffer_capacity and
+ * decrypts those buffers and returns a list of decrypted buffers in the decrypted_buffs list.
+ *
+ * @param tls - The qd_tls_t object which contains the proton tls session information
+ * @param pn_raw_conn - raw connection from which were going to take buffers from
+ * @param decrypted_buffs - A pointer to qd_adaptor_buffer_list which will contain the decrypted buffs. Make sure this
+ * list is initialized and empty.
+ *
+ * @return QD_TLS_ERROR if there is an error when trying to decrypt an encrypted buffer, the number of undecrypted bytes
+ * totally processed otherwise
+ *
+ */
+int qd_tls_decrypt(qd_tls_t                 *tls,
+                   pn_raw_connection_t      *pn_raw_conn,
+                   qd_adaptor_buffer_list_t *decrypted_buffs,
+                   qd_adaptor_buffer_list_t *granted_read_buffs);
```

Comment:
Why does this function take a list of buffers that were already granted to the pn_raw_connection_t?  Why not simply take available full read buffers from the raw connection?